### PR TITLE
feat(policy): add LstmBurnPolicy recurrent actor-critic with seeded LSTM init

### LINKED ISSUE #285: feat(policy): LstmBurnPolicy + seeded LSTM init
Phase 1 of the recurrent-policy epic #262. See design note [`docs/RECURRENT_POLICY_DESIGN.md`](../blob/main/docs/RECURRENT_POLICY_DESIGN.md).

## Scope
Implement `LstmBurnPolicy<B>` in `src/policy/lstm.rs` alongside the concrete `MlpBurnPolicy` — no formal `Policy` trait exists; each policy is a concrete `#[derive(Module)]` struct.

## Deliverables
- LSTM trunk (Burn 0.21 `Lstm<B>`, rank-3 forward, `Option<LstmState<B, 2>>`) + policy head (logits) + value head.
- `forward_step(obs, state) -> (logits, value, new_state)` for single-step rollout collection (threads `(h, c)` in/out).
- `evaluate_sequences(obs_seq, actions, initial_state, episode_starts) -> (log_probs, entropy, values)` for the rank-3 training forward, resetting `(h, c) -> 0` at **episode boundaries** where `episode_starts = terminated || truncated`. The `episode_starts` parameter carries per-step episode-boundary (done) flags, **not** terminated-only: Thrust's rollout loop resets envs on `terminated || truncated` (`examples/games/cartpole/train_cartpole_modern.rs:204,214`) and CartPole/MaskedCartPole truncate at `max_steps = 500` (`src/env/games/cartpole.rs:105`), so a truncated step ends the episode and its state must reset. This is distinct from GAE, which bootstraps on `terminated` only (see the GAE-vs-state-reset note in the design doc's Q2).
- `LstmBurnConfig { hidden_dim, seed: Option<u64>, .. }`. When `seed` is `Some`, override all **eight** gate `Linear` layers (4 `GateController`s x {input_transform, hidden_transform}) via the existing `seeded_layer_weights` / `derive_layer_seed` pattern from `src/policy/mlp.rs:100-125`, drawing from `seeded_init.rs`. This bypasses Burn's unseeded default `XavierNormal` init.
- Optional GRU variant may be deferred.

## Acceptance criteria
- [ ] `LstmBurnPolicy<B>` derives Burn `Module` + `AutodiffModule`.
- [ ] `forward_step` returns correct state shapes `[batch, hidden_dim]`.
- [ ] Two constructions with the same `seed` produce bit-identical weights (mirror `test_with_seed_is_bit_identical_orthogonal` in `mlp.rs`).
- [ ] Hidden state resets to zeros at episode boundaries (`terminated || truncated`), carries forward otherwise. (Truncation ends the episode too — resetting on terminated alone would leak stale state into the fresh post-truncation episode.)
- [ ] Registered under `#[cfg(feature = "training")]` in `src/policy/mod.rs`.

Depends on: nothing. Blocks: Phase 2, Phase 3.

Refs #262

### PR DESCRIPTION
**Pull request by rjwalters**
## Summary

Phase 1 of the recurrent-policy epic (#262). Adds `LstmBurnPolicy<B>` in `src/policy/lstm.rs` — a single-layer LSTM actor-critic built as a concrete Burn `Module` (mirroring `MlpBurnPolicy`; there is no formal `Policy` trait). It wraps Burn 0.21's rank-3 stateful `Lstm` trunk with shared policy/value heads, and closes the reproducibility gap by seeding the LSTM's otherwise-unseedable gate init.

Scope is **policy only** — no `RolloutBuffer`, trainer, or POMDP env changes (those are Phases 2/3, #286/#287).

## Changes

- **`src/policy/lstm.rs`** (new):
  - `LstmBurnPolicy<B>` — `#[derive(Module, Debug)]`; fields `lstm: Lstm<B>`, `policy_head`, `value_head`. Deriving `Module` transitively provides the `AutodiffModule` impl the training path needs.
  - `forward_step(obs, state) -> (logits [batch, A], value [batch], LstmState [batch, hidden])` — single-step rollout inference, threads `(h, c)` in/out.
  - `evaluate_sequences(obs_seq [n_env, T, obs], actions [n_env, T], initial_state, episode_starts [n_env, T]) -> (log_probs, entropy, values)`, each `[n_env, T]` — the rank-3 training forward. Steps the LSTM across `T`, zeroing `(h, c)` **per-env** at episode boundaries via a `1 - episode_starts[:, t]` broadcast mask. `episode_starts` carries `terminated || truncated` (done), the deliberate asymmetry with GAE's terminated-only bootstrap preserved from the design note's Q2.
  - `LstmBurnConfig { hidden_dim, use_orthogonal_init, seed }` + `with_seed`. Seeded path overrides all **8** gate `Linear`s (4 `GateController`s × {input,hidden}_transform) plus both heads from `StdRng`-driven buffers via the existing `seeded_layer_weights` / `derive_layer_seed` helpers (layer indices 0–9), bypassing Burn's unseedable `XavierNormal`.
- **`src/policy/mod.rs`**: register `pub mod lstm;` under `#[cfg(feature = "training")]`.

## Acceptance Criteria Verification

| Criterion | Status | Verification |
|-----------|--------|--------------|
| Derives `Module` + `AutodiffModule` | ✅ | `#[derive(Module, Debug)]`; tests use `Autodiff<NdArray<f32>>` backend |
| `forward_step` returns state shapes `[batch, hidden_dim]` | ✅ | `test_forward_step_shapes` asserts cell/hidden dims `[1, 8]` |
| Same seed ⇒ bit-identical weights | ✅ | `test_with_seed_is_bit_identical` (mirrors `mlp.rs:758`); `test_seeded_gates_are_decorrelated` guards per-layer seed derivation |
| Hidden state resets at `terminated \|\| truncated`, carries otherwise | ✅ | `test_evaluate_sequences_episode_boundary_reset`: step-1 value under reset equals a fresh zero-state `forward_step`; without reset it differs |
| Registered under `#[cfg(feature = "training")]` | ✅ | `src/policy/mod.rs` |

## Test Plan

- `cargo test --lib --features training policy::lstm` — 6/6 pass (unseeded construction, forward_step shapes, evaluate_sequences shapes, bit-identical seed, gate decorrelation, episode-boundary reset).
- `cargo clippy --lib --features training` and `--all-targets` — clean, no warnings.
- `cargo fmt` — applied; only the two in-scope files changed.

Closes #285
Refs #262

### CODE REVIEW

**Comment by rjwalters** [comment]
## Changes Requested — Documentation CI failing

The implementation is **substantively correct and I want to approve it** — the only blocker is a hard Documentation CI failure. Once that is fixed this should sail through.

### Blocking: broken intra-doc links (Documentation job fails)

The `Documentation` check runs `RUSTDOCFLAGS="-D warnings" cargo doc --no-deps --all-features`, and `-D warnings` promotes `rustdoc::broken_intra_doc_links` to a hard error. I reproduced it locally with just `--features training` (so it is **not** an all-features artifact):

```
error: unresolved link to `LstmBurnPolicy`      (= note: no item named `LstmBurnPolicy` in scope)
error: unresolved link to `Lstm`
error: unresolved link to `LstmBurnPolicy::forward_step`
error: unresolved link to `LstmBurnPolicy::evaluate_sequences`
error: unresolved link to `LstmConfig`
error: unresolved link to `seeded_layer_weights`
error: unresolved link to `derive_layer_seed`
error: could not document `thrust-rl`
```

All seven are the **short-name** ``[`Name`]`` links in the module-level `//!` header of `src/policy/lstm.rs` (lines 1–38). Every *fully-qualified* link in that same header (e.g. `` [`crate::policy::mlp::MlpBurnPolicy`] ``, `` [`burn::nn::GateController`] ``, `` [`StdRng`](rand::rngs::StdRng) ``) resolves fine — the short names do not resolve in the `//!` context.

**Fix:** fully-qualify (or drop the link brackets on) the seven failing references, e.g.
- `` [`LstmBurnPolicy`] `` → `` [`LstmBurnPolicy`](crate::policy::lstm::LstmBurnPolicy) `` (and the two method links similarly)
- `` [`Lstm`] `` / `` [`LstmConfig`] `` → `` [`Lstm`](burn::nn::Lstm) `` / `` [`LstmConfig`](burn::nn::LstmConfig) ``
- `` [`seeded_layer_weights`] `` / `` [`derive_layer_seed`] `` → these are `pub(crate)` in `mlp.rs`, so a link from this public module's docs may additionally trip `rustdoc::private_intra_doc_links` under `-D warnings`. Safest to drop the brackets and leave them as plain `` `seeded_layer_weights` `` code spans, or link the module `` [`crate::policy::mlp`] `` instead.

Then confirm clean with:
```
RUSTDOCFLAGS="-D warnings" cargo doc --no-deps --all-features
```

### Everything else checks out (for the record)

- **`evaluate_sequences` masking** — `keep = 1 - episode_starts[:, t]` (`[n_env, 1]`) multiplies `(cell, hidden)` *before* the step at `t`, broadcasting over `hidden`. Zeroes state at episode starts exactly per design-note Q2. Shape flow and the categorical log-softmax/gather/entropy over the action axis are correct.
- **Reset test** genuinely proves *conditional* reset: step-1 value under reset equals a fresh zero-state `forward_step`, and the no-reset case differs — both branches exercised.
- **Seeded init** — all 8 gate Linears (idx 0–7) + both heads (idx 8,9) overwritten with distinct `derive_layer_seed` indices; no collision, none missed; bias deterministically zeroed via `linear_from_weights`. Bit-identity + decorrelation tests pass.
- **Scope** — policy only (`lstm.rs` + `mod.rs` registration under `#[cfg(feature = "training")]`); no buffer/trainer/env changes.
- **Design-note conformance** — `episode_starts = terminated || truncated`, GAE asymmetry documented and preserved.
- Local `cargo test --lib --features training policy::lstm`: **6/6 pass**. Clippy + Rustfmt CI: pass.

Fix the doc links, push, and re-request review — this is otherwise ready.

**Comment by rjwalters** [comment]
Documentation CI fix pushed.

Fixed the seven broken intra-doc links in the \`//!\` module header of \`src/policy/lstm.rs\`:

- \`LstmBurnPolicy\`, \`LstmBurnPolicy::forward_step\`, \`LstmBurnPolicy::evaluate_sequences\` -> fully-qualified with \`crate::policy::lstm::...\`
- \`Lstm\`, \`LstmConfig\` -> fully-qualified with \`burn::nn::...\`
- \`seeded_layer_weights\`, \`derive_layer_seed\` (pub(crate) in \`mlp.rs\`) -> dropped the link brackets to plain code spans, avoiding \`rustdoc::private_intra_doc_links\` under \`-D warnings\`.

No code semantics changed — doc header only, exactly as requested.

Verified locally clean:
- \`RUSTDOCFLAGS="-D warnings" cargo doc --no-deps --features training\` -> builds with no unresolved-link errors
- \`cargo fmt --check\` -> clean

Ready for re-review.

**Comment by rjwalters** [comment]
✅ **Approved!** (re-review after Doctor pass)

This re-review is scoped to the Documentation CI regression that was the sole blocker last pass. My predecessor already cleared all substantive code: `evaluate_sequences` per-env boundary masking, seeded init across all 10 layers, policy-only scope, design-note (RECURRENT_POLICY_DESIGN.md Q2/Q4) conformance, and 6/6 tests.

**Diff since previous review (Doctor commit 9d0ea04) — doc header only, no semantic drift:**
- The only touched file is `src/policy/lstm.rs`, and only the `//!` module header (12 insertions / 11 deletions).
- Five intra-doc links fully-qualified: `LstmBurnPolicy`, `forward_step`, `evaluate_sequences` → `crate::policy::lstm::…`; `Lstm`, `LstmConfig` → `burn::nn::…`.
- The two `pub(crate)` helper references (`seeded_layer_weights` / `derive_layer_seed`) converted to plain code spans, correctly avoiding `rustdoc::private_intra_doc_links`.
- No code, signatures, or test behavior changed.

**CI — all 9 checks green (run 28724424083):**
- Documentation — pass (1m32s) ← previously failing, now resolved
- Tests (Linux + Burn NdArray) — pass (10m52s) ← pending last review
- Tests (macOS + Burn NdArray) — pass (7m41s) ← pending last review
- Clippy — pass; Rustfmt — pass; Security Audit — pass; Check (no training) macOS + Ubuntu — pass; Publish dry-run — pass

Merge state: **CLEAN**. PR closes #285 via proper auto-close syntax. Ready to merge.
